arch/cortex-m/src/lib.rs

```diff
@@ -210,52 +210,9 @@ pub unsafe extern "C" fn unhandled_interrupt() {
     panic!("Unhandled Interrupt. ISR {} is active.", interrupt_number);
 }
 
-/// Assembly function called from `UserspaceKernelBoundary` to switch to an
-/// an application. This handles storing and restoring application state before
-/// and after the switch.
-#[cfg(all(target_arch = "arm", target_os = "none"))]
```

Comment:
From `rustc --print target-list`

Can this `cfg` maybe be an `any` of 

```
thumbv7a-pc-windows-msvc
thumbv7a-uwp-windows-msvc
thumbv7em-none-eabi
thumbv7em-none-eabihf
thumbv7m-none-eabi
thumbv7neon-linux-androideabi
thumbv7neon-unknown-linux-gnueabihf
thumbv7neon-unknown-linux-musleabihf
thumbv8m.base-none-eabi
thumbv8m.main-none-eabi
thumbv8m.main-none-eabihf
```

(or possibly not even those last three)?

Reply:
We tried `#[cfg(all(target_arch = "thumbv7em", target_os = "none"))]` and it ignores it when we compile for the m4 (TARGET=thumbv7em-none-eabi)

Reply:
`target_arch` is `arm` here, not `thumbv7em` (it is different from the first part of the target triple).

I haven't looked at this PR to see what you are trying to achieve, but I don't think you can access a target triple from `cfg`. You can access it from `build.rs` (as the `CARGO_BUILD_TARGET` environment flag) if you really need to know.

Reply:
I don't think using build.rs is useful, we need to make the compiler to ignore a function that contains armv7 code (thumb2).

Reply:
yeah, I think if we wanted to use `cfg` we would need to create a custom attribute which is set differently for thumbv6 boards than for thumbv7. I imagine 4 possible solutions here:

1. custom cfg options for thumbv7 vs thumbv6, with the correct cfg set in build.rs for each cortex-m board.
2. Avoid the problem for now by replacing `ldmia` and `stdmia` with multiple instructions that accomplish the same thing (I think this is possible but didn't look deeply into it)
3. move the function to cortex-m3 and import from m4 (I do not like this, as the m4 crate should not depend on the m3 crate)
4. Create `thumbv6m` and `thumbv7m` crates and integrate them into the existing hierarchy somehow

I probably prefer (2) for now but am curious what others think

Reply:
It seems like we need to fix rust's `target_arch` cfg for them, as "arm" is insufficient (arg!!). Could we define `target_arch_tock` in our build system to be more specific?

Reply:
That is, I prefer (1) (since that fixes the underlying problem that rust has created, and might be useful in the future since we should be able to cfg on an _actual_ architecture, not just a family of architectures. Perhaps rust will fix this in the future.).

I guess we could do (4), but I find "cortex-mX" more user friendly than "thumbvXX" since humans are exposed to marketing names, for better or worse. I think we could balance this with yet-more-crates, but that fragments code even more, and makes porting fixes to all relevant platforms more challenging.

Reply:
It looks like rust does give us a way to do this already: https://github.com/tock/tock/pull/2452

Reply:
We are now more pedantic about exactly which arm architecture, so this problem should be avoided now that #2452 has been merged.